uniffi_bindgen/src/interface/types.rs

```diff
@@ -128,243 +132,349 @@ impl From<&Type> for FFIType {
-                return ci.add_type_definition(self.identifier.0, Type::Error(name));
+                return types.add_type_definition(self.identifier.0, Type::Error(name));
             }
         }
-        let name = self.identifier.0.to_string();
-        if resolve_builtin_type(&self.identifier).is_some() {
-            bail!("please don't shadow builtin types (enum {})", name);
-        }
-        ci.add_type_definition(self.identifier.0, Type::Enum(name))
+        types.add_type_definition(self.identifier.0, Type::Enum(name))
     }
 }
 
 impl TypeFinder for weedle::TypedefDefinition<'_> {
-    fn find_type_definitions(&self, ci: &mut ComponentInterface) -> Result<()> {
+    fn add_type_definitions_to(&self, types: &mut TypeUniverse) -> Result<()> {
         if self.attributes.is_some() {
             bail!("no typedef attributes are currently supported");
         }
-        match resolve_builtin_type(&self.identifier) {
-            Some(_) => bail!(
-                "please don't try to redefine builtin types, kthxbye ({:?})",
-                self
-            ),
-            None => {
-                // For now, we assume that the typedef must refer to an already-defined type, which means
-                // we can look it up in the ComponentInterface. This should suffice for our needs for
-                // a good long while before we consider implementing a more complex delayed resolution strategy.
-                ci.add_type_definition(self.identifier.0, self.type_.resolve_type_definition(ci)?)
-            }
-        }
+        // For now, we assume that the typedef must refer to an already-defined type, which means
+        // we can look it up in the TypeUniverse. This should suffice for our needs for
+        // a good long while before we consider implementing a more complex delayed resolution strategy.
+        let t = types.resolve_type_expression(&self.type_)?;
+        types.add_type_definition(self.identifier.0, t)
     }
 }
 
-/// Trait to help resolving an IDL type node to a `HighLevelType`.
+/// Trait to help resolving an IDL type node to a [Type].
 ///
 /// Ths trait does structural matching against type-related weedle AST nodes from
-/// a parsed IDL file, turning them into a corresponding `HighLevelType`. It uses the
-/// mapping from names to types previous built by the `TypeFinder` trait in order to
-/// resolve types by name.
+/// a parsed IDL file, turning them into a corresponding [Type] struct. It uses the
+/// known type definitions in a [TypeUniverse] to resolve names, and so it assumes
+/// that we've already done a [TypeFinder] pass.
 ///
 /// (And to be honest, a big part of its job is to error out if we encounter any of the
 /// many, many WebIDL type definitions that are not supported by uniffi.)
 pub(crate) trait TypeResolver {
-    fn resolve_type_definition(&self, ci: &ComponentInterface) -> Result<Type>;
+    fn resolve_type_expression(&self, types: &mut TypeUniverse) -> Result<Type>;
```

Comment:
The thing this method operates on is not a "definition" in any meaningful sense - it's an expression for a type, built up from the named types we discovered in the type-finding pass. So, I've renamed the method.

Reply:
I'm still not wild about the name, (either the `resolve` or the `expression` bits).

This might be that the difference in naming between the `TypeResolver` and `TypeFinder` is so small.

Reply:
That's fair; I think the new name is *better*, but that's not the same thing as *good*.

"Resolve" is probably not what we're really mean here. Not for this PR, but I'd be happy to try to brainstorm something better. Maybe something using "builder" could be helpful, since this builds more complex structured types out of the named primitives we found in the previous pass?

Good naming is hard :-/